LUCENE-7785: Move dictionary for Ukrainan analyzer to external dependency

arysin opened the pull request.

arysin commented:
It would be also nice to merge this into 6x branch

dweiss commented on lucene/analysis/morfologik/src/java/org/apache/lucene/analysis/uk/UkrainianMorfologikAnalyzer.java:
```diff
@@ -107,11 +107,18 @@ public UkrainianMorfologikAnalyzer(CharArraySet stopwords, CharArraySet stemExcl
     builder.add("\u0301", "");
-    NormalizeCharMap normMap = builder.build();
+    builder.add("\u00AD", "");
+    builder.add("\uFEFF", "");
```
byte order mark shouldn't be replaced to nothing... if you have a byte order mark in your character input (reader) then your conversion from bytes is screwed up somewhere before.

Trey314159 replied:
Just stumbled across this dicussion today. I'm the Wikimedia guy that arysin linked to above. And while it is _waaaaay_ to late, I just wanted to point out that while U+FEFF is used as the byte order mark at the beginning of a text stream, it is [also used](https://en.wikipedia.org/wiki/Byte_order_mark#Usage) as a "zero-width non-breaking space" within a text. That use was deprecated as of Unicode 3.2, but it is still quite common in practice on lots of Wikipedias. ICU normalization converts it to the empty string, which is how it usually gets handled on the wikis where we have had the chance to customize the analyzers, though that requires being able to open them up, so to speak.

dweiss commented on lucene/analysis/morfologik/src/java/org/apache/lucene/analysis/uk/UkrainianMorfologikAnalyzer.java:
```diff
@@ -145,7 +152,7 @@ protected TokenStreamComponents createComponents(String fieldName) {
   private static Dictionary getDictionary() {
     try {
-      return Dictionary.read(UkrainianMorfologikAnalyzer.class.getResource("ukrainian.dict"));
+      return Dictionary.read(UkrainianMorfologikAnalyzer.class.getResource("/ua/net/nlp/ukrainian.dict"));
```
This shouldn't be a class-relative read looking for an absolute resource. Rather, I think, it should be classloader-relative UkrainianMorfologikAnalyzer.class.getClassLoader().getResource("ua/net/nlp/ukrainian.dict").

arysin replied:
I am a bit confused on why for class loader getResource() we need relative path (surprisingly the full one starting with "/" does not work for me), but otherwise it makes sense so I'll push the fix shortly.

sarowe commented on lucene/analysis/morfologik/ivy.xml:
```diff
@@ -25,6 +25,7 @@
     <dependency org="org.carrot2" name="morfologik-polish" rev="${/org.carrot2/morfologik-polish}" conf="compile"/>
     <dependency org="org.carrot2" name="morfologik-fsa" rev="${/org.carrot2/morfologik-fsa}" conf="compile"/>
     <dependency org="org.carrot2" name="morfologik-stemming" rev="${/org.carrot2/morfologik-stemming}" conf="compile"/>
+    <dependency org="ua.net.nlp" name="morfologik-ukrainian-search" rev="${ua.net.nlp.morfologik-ukrainian-search.version}" conf="compile"/>
```
The `rev` attribute value has to be of the form `"${/org/name}"` - this will not pass precommit.  You have to make such an entry in `ivy-versions.properties` - look at all the other examples there.

sarowe commented on lucene/ivy-versions.properties:
```diff
@@ -203,6 +203,8 @@ org.carrot2.morfologik.version = 2.1.1
 /org.carrot2/morfologik-polish = ${org.carrot2.morfologik.version}
 /org.carrot2/morfologik-stemming = ${org.carrot2.morfologik.version}
 
+ua.net.nlp.morfologik-ukrainian-search.version = 3.7.4
```
This is fine, but you're missing the `/org/name=version` thing that's required to be used in `ivy.xml` files.

If you look at the rest of this file, you can see that some `/org/name` keys share the same version, and so a variable is introduced to simplify maintenance.  In your case, when there is only one `/org/name` key using this version, you should IMO eliminate the `ua.net.nlp.morfologik-ukrainian-search.version` variable entirely and add `/ua.net.nlp/morfologik-ukrainian-search = 3.7.4`.

arysin commented on lucene/analysis/morfologik/src/java/org/apache/lucene/analysis/uk/UkrainianMorfologikAnalyzer.java:
```diff
@@ -107,11 +107,18 @@ public UkrainianMorfologikAnalyzer(CharArraySet stopwords, CharArraySet stemExcl
     builder.add("\u0301", "");
-    NormalizeCharMap normMap = builder.build();
+    builder.add("\u00AD", "");
+    builder.add("\uFEFF", "");
```
That was from the note [Wikimedia guys suggested](https://www.mediawiki.org/wiki/User:TJones_(WMF)/Notes/Ukrainian_Morfologik_Analysis#Recommendations_.26_Plan), but agree it does not make sense here, I'll remove it

arysin commented:
Thanks for the prompt feedback guys, I think I've addressed all the issues with the latest commits but please let me know if something is still wrong.
Once we're good with master branch I'll prepare the 6x changes.

dweiss commented on lucene/licenses/morfologik-ukrainian-search-NOTICE.txt:
```diff
@@ -0,0 +1,6 @@
+morfologik-ukrainian-search is a POS tag dictionary in morfolgik format adjusted for searching.
```
typo in morfolgik

dweiss commented on lucene/licenses/morfologik-ukrainian-search-NOTICE.txt:
```diff
@@ -0,0 +1,6 @@
+
+Note: to better fit into full-text search model this dictionary has all word forms in lower case but keeps lemmas for proper nouns in upper case.
+
+Licensed under GPL/LGPL, CC BY-NC-SA 4.0, and Apache License 2.0.
```
Hmm; not a lawyer, but this should better state "Licensed under Apache License 2.0"... Those multiple licenses don't agree with each other -- at least it should state *or*, not *and*.
